ENH: Refactor module

Eric89GXL opened the pull request:
Closes #34.

This refactors the module as discussed in the above issue.

Eric89GXL commented on api/util.py:
```diff
@@ -80,6 +83,17 @@ def commit_file(dbc, _id, datainfo, filepath, data_path, force=False):
+        fmt = logging.Formatter(fmt=log_fmt, datefmt=datefmt)
+        hdlr.setFormatter(fmt)
+        hdlr.setLevel(logging.DEBUG)
+        log.addHandler(hdlr)
```
This is the thing that made debugging much easier, namely logging the `commit_file` calls to `persistent/logs/commit_file.log`. It's a bit hackishly coded right now, I suspect it will be easier to get this right with a single-repo setup.

gsfr replied:
I don't like this at all. Let's get rid of it. How is this better than a ```tail -f | grep``` of the UWSGI log, anyway?

Also, the logging module has ways of selectively using handlers for certain messages. Not that I'm saying we should do that, though.

Eric89GXL replied:
I agree it's a hack and I don't like it either, but none of these logging
messages got saved currently as far as I could find. The first thing I
checked was uwsgi and didn't see anything, and I think I actually grep
-R'ed to check and didn't find the messages anywhere. I think it's better
to have the hack in place, working but ugly, than not have the output at
all, but I'll get rid of it if you want. Ideally we could do it in a more
principled way, but I don't get how the pieces for together well enough yet
to see what that is...

gsfr replied:
During bootstrapping, messages only go to ```stdout```, which you can probably redirect if you really want it in a file. Reaper uploads should surely be logged to ```uwsgi.log```. If they're not, that needs to be fixed. Might be as simple as adjusting the log level down.

Eric89GXL commented:
Requires https://github.com/scitran/scitran/pull/64 for `scitran` to work properly

gsfr commented:
Thanks, Eric. Very cool.

A couple of things:
- Shouldn't we create the ```scitran.api``` namespace?
- ```bin/centralclient.py``` should be executable, not ```api/centralclient.py```.
- Isn't a ```data``` directory for a single file a bit excessive? I, for one, find it plain annoying.
- Please resolve the merge conflicts.

Eric89GXL commented:
> Shouldn't we create the scitran.api namespace?

Yeah, but unfortunately I don't think we can right now. IIRC there are a couple of things that need to import from both `scitran.api` and `scitran.data`, but they aren't actually both nested under a single `scitran` folder currently. So anything that tries to import both will only find one of them, and not the other (which it finds would be dependent upon `sys.path` order). But it shouldn't be difficult to nest `api` under `scitran` in the eventual "big merge" situation once this is in place.

> bin/centralclient.py should be executable, not api/centralclient.py.

Whoops! Yeah I `git mv`'ed the latter, and created the former without changing perms. Fixed with latest commit.

> Isn't a data directory for a single file a bit excessive? I, for one, find it plain annoying.

It's overkill for now. I usually do that in case the number of data files grows -- then it's convenient to provide some structure. But for now it's overkill and it can always be created later, so I've removed it.

> Please resolve the merge conflicts.

Done.

I deleted my `persistent` folder and re-reaped some MEEG data, and it still works. So I don't think I broke anything with the changes :)

gsfr commented:
Cool. Makes sense about the namespace. Unfortunate, though.

If you guys are interested (@Eric89GXL, @kofalt), I'd like to do a Hangout next week to discuss the bigger refactoring plans.

I will be traveling to MSP tomorrow, but should have time to test this from scratch at the airport.

@kofalt Is ```apps.py``` still relevant?

@Eric89GXL Can we now safely rename ```collections_.py``` to ```collections.py```?

Eric89GXL commented:
Yeah I can do the renaming tomorrow morning.

Feel free to pick some times that work for you all for hangout. My schedule
is pretty flexible, but I'm on eastern time right now so before 2pm pacific
is better for me but not required.

Eric89GXL commented on api/apps.py:
```diff
@@ -107,8 +104,11 @@ def post(self):
             log.debug('Recieved App: %s' % app_meta.get('_id'))
+        """
+        # XXX util.insert_app doesn't exist...?
+        raise NotImplementedError
```
BTW just wanted to point this out -- this method made use of a nonexistent `util.insert_app` function, so I don't think it is ever actually called. Should I just delete it? Right now I hackishly "commented" it out by making the code a big string.

Eric89GXL commented:
`collections_` moved.

@gsfr don't forget you need scitran/scitran#64 simultaneously when you go to test.

gsfr commented:
Works!

- Not sure why we now have a Makefile to clean stuff.
- Wondering if ```api.wsgi``` should be in ```bin```. It's basically an executable python file. Is the currently layout typical for WSGI apps?
- I don't think we actually have a need for ```bin/centralclient.py```. We always use UWSGI.
- @kofalt Is ```apps.py``` still relevant? (Restating for completeness, not pressure.)

Eric89GXL commented:
> Not sure why we now have a Makefile to clean stuff.

The `Makefile` is there because it's nice to be able to do `make clean` to ensure all the `.pyc` files have been cleared out, or their Python3 equivalents if eventually this moves to that platform. I can remove it if it seems unnecessary. Usually other commands end up in there, like `make test` or `make flake` (I guess `make lint` in this case probably), but maybe this will end up being redundant with `live.sh` somehow.

> Wondering if api.wsgi should be in bin. It's basically an executable python file. Is the currently layout typical for WSGI apps?

I have never worked with it, so it doesn't really matter to me. If it's primarily used as an executable, which it sounds like it is, it's probably worth moving.

> I don't think we actually have a need for bin/centralclient.py. We always use UWSGI.

I wasn't sure what that was used for, so I'm happy to remove it. It's just copy-pasted from some code that was already in there, perhaps it was outdated.

gsfr commented:
>> Not sure why we now have a Makefile to clean stuff.

> The Makefile is there because it's nice to be able to do make clean to ensure all the .pyc files have been cleared out, or their Python3 equivalents if eventually this moves to that platform. I can remove it if it seems unnecessary. Usually other commands end up in there, like make test or make flake (I guess make lint in this case probably), but maybe this will end up being redundant with live.sh somehow.

I'd say we keep it lean until a real need arises.

>> Wondering if api.wsgi should be in bin. It's basically an executable python file. Is the currently layout typical for WSGI apps?

> I have never worked with it, so it doesn't really matter to me. If it's primarily used as an executable, which it sounds like it is, it's probably worth moving.

Agreed. This will very likely need a matching change in scitran/scitran#64.

>> I don't think we actually have a need for bin/centralclient.py. We always use UWSGI.

> I wasn't sure what that was used for, so I'm happy to remove it. It's just copy-pasted from some code that was already in there, perhaps it was outdated.

Yeah, it's legacy. Let's ditch it.

Eric89GXL commented:
@gsfr comments addressed with changes here and in https://github.com/scitran/scitran/pull/64

Eric89GXL commented:
I re-tested locally and things still appear to be okay

gsfr commented:
I also just retested. Everything still works. I'll be adding a few line comments shortly.

gsfr commented on api/util.py:
```diff
@@ -15,9 +12,15 @@
+datefmt = '%Y-%m-%d %H:%M:%S'
+log = logging.getLogger('scitran.api')
+logging.basicConfig(format=log_fmt, datefmt=datefmt, level=logging.INFO)
+
```
Ideally, we'd only have one definition of our logging style. We currently have this one and one in ```api.wsgi```. ```util.py``` is probably the right place, but I prefer the definition we have in ```api.wsgi```.

gsfr commented on api/util.py:
```diff
@@ -150,6 +168,12 @@ def _update_db(db, datainfo):
+    except KeyError:  # this can happen if nims_metadata_status == None
+        session_label = '(unknown)'
+    log.info('Using group_id="%s", project_name="%s", and session_label="%s"'
+             % (project['group'], project['name'], session_label))
```
Let's try to keep the overall number of log lines per sorted file to a minimum. I'm currently seeing 3 lines per file at ```info``` level, plus one for parsing. One line at ```info``` and one more at ```debug``` seems like a good target. I see the issue of having access to the relevant info at the right time without returning a bunch of basically useless info.

I would play with it myself, but I won't because I can't commit into your PR. If you're going to continue contributing at this rate, it might make sense to make you a commiter for our repos. Thoughts?

Please also try to follow the layout of existing log entires. Note the spacing.

Also, this should be a single line.

Eric89GXL replied:
You can always open a PR into my branch with changes, it might work better
than silently committing. I could also give you write permissions to my
fork if you'd prefer to avoid opening PRs.

gsfr replied:
I'm worried about things getting too confusing with PRs into PRs.

gsfr commented on api/util.py:
```diff
@@ -80,6 +85,17 @@ def commit_file(dbc, _id, datainfo, filepath, data_path, force=False):
+        log.addHandler(hdlr)
+    else:
+        hdlr = None
     log.info('Sorting     %s' % filename)
```
```bootstrap.py:87``` is redundant with this. This maybe useful for reaper uploads, though, and we might need an extra line in ```reaper()``` in ```core.py```.

Eric89GXL commented:
@gsfr comments addressed. You were right about the log, just setting `log.setLevel(logging.INFO)` made the messages appear in the `uwsgi` log.

kofalt commented:
On review this and scitran/scitran#64 make sense to me, what's the status here?
@gsfr Do we want to do a hangout discussing plans?

I'm ready to sanity check + merge these two PRs.

Eric89GXL commented:
Ready from my end, but I don't think @gsfr has had a chance to look at the changes I made to address his comments.

gsfr commented:
Thanks for addressing these final things, @Eric89GXL. I'm good to go on this, bar the one line comment I just made.

@kofalt A final sanity check from scratch would be great.

@kofalt Verdict on relevance of ```apps.py```?

Eric89GXL commented:
@gsfr done

kofalt commented:
Nuke apps.py. But I wouldn't make that a requirement of this PR. Eric feel free to delete :)

Eric89GXL commented:
@kofalt @gsfr done, it required removing the corresponding entry from the routes in `api.py`, I hope that's what you had in mind. If not, I can remove the last commit and it can be done properly in a subsequent PR.

@kofalt did the final sanity check from scratch work for you? It runs fine over here.

gsfr commented:
I did another test from scratch this morning, including scitran/data#28 and scitran/scitran#64. Everything seems to have worked.

Unfortunately, we have a merge conflict now. Please rebase or merge and then we are finally ready to go.

Eric89GXL commented:
@gsfr done, I'll re-run it again just to make sure it still works

kofalt commented:
I'll look at this within the hour :)

Eric89GXL commented:
Sample data loading and MEEG reaping both appear to work fine over here

kofalt commented:
Sanity checks passed; merging both PRs.
